src/btree/bt_curnext.c

```diff
@@ -980,7 +980,8 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 F_CLR(cursor, WT_CURSTD_BOUND_LOWER_INCLUSIVE);
                 ret = __wt_compare_bounds(
                   session, cursor, &cbt->iface.key, cbt->recno, false, &key_out_of_bounds);
-                WT_ASSERT(session, ret == 0 && !key_out_of_bounds);
+                WT_ASSERT_ALWAYS(session, ret == 0 && !key_out_of_bounds,
+                  "Failed to verify cursor position after walking. Cursor is out of bounds");
```

Comment:
Perhaps: "Bounded cursor logic resulted in the cursor being an illegal state." This assertion was intentionally in a #HAVE_DIAGNOSTIC block because it validates prior logic but is somewhat expensive. If this assertion fails explaining why that is bad is hard so I think a specific message won't be very valuable.

Reply:
[4ec1860](https://github.com/wiredtiger/wiredtiger/pull/8623/commits/4ec1860657e6aeef19c7f7a20fa12817c7c791a1)

Reply:
@luke-pearson the `#HAVE_DIAGNOSTIC` block has been replaced with a runtime `DIAG_ASSERTS_ENABLED` check so we still skip these checks unless the user sets a flag on the connection. The change from `ASSERT` to `ASSERT_ALWAYS` is so this code will actually run when we turn on extra asserts at runtime.

r.e. `explaining why that is bad is hard` would it be useful to turn this on in production? If not we can change this back to a `#HAVE_DIAGNOSTIC` block which is used in testing only.

Reply:
This assertion makes sense as a runtime enableable assertion. I assume if a user was turning that on it'd likely be at our request so we could parse the message. That's why I don't think we need a lot of detail here.